Source/Windows/ARM64EC/Module.cpp

```diff
@@ -265,43 +266,74 @@ void InitSyscalls() {
   if (LoadConfig->VolatileMetadataPointer < Address || LoadConfig->VolatileMetadataPointer + sizeof(IMAGE_VOLATILE_METADATA) >= EndAddress) {
-    return;
+    goto no_file_metadata;
   }
 
-  const auto* VolatileMetadata = reinterpret_cast<IMAGE_VOLATILE_METADATA*>(LoadConfig->VolatileMetadataPointer);
-  if (!VolatileMetadata || Address + VolatileMetadata->VolatileAccessTable + VolatileMetadata->VolatileAccessTableSize >= EndAddress ||
-      Address + VolatileMetadata->VolatileInfoRangeTable + VolatileMetadata->VolatileInfoRangeTableSize >= EndAddress) {
-    return;
-  }
+  {
+    const auto* VolatileMetadata = reinterpret_cast<IMAGE_VOLATILE_METADATA*>(LoadConfig->VolatileMetadataPointer);
+    if (!VolatileMetadata || Address + VolatileMetadata->VolatileAccessTable + VolatileMetadata->VolatileAccessTableSize >= EndAddress ||
+        Address + VolatileMetadata->VolatileInfoRangeTable + VolatileMetadata->VolatileInfoRangeTableSize >= EndAddress) {
+      goto no_file_metadata;
+    }
 
-  fextl::set<uint64_t> VolatileInstructions;
-  const auto* VolatileAccessTableBegin = reinterpret_cast<IMAGE_VOLATILE_RVA_METADATA*>(Address + VolatileMetadata->VolatileAccessTable);
-  const auto* VolatileAccessTableEnd =
-    VolatileAccessTableBegin + (VolatileMetadata->VolatileAccessTableSize / sizeof(IMAGE_VOLATILE_RVA_METADATA));
-  for (auto It = VolatileAccessTableBegin; It != VolatileAccessTableEnd; It++) {
-    VolatileInstructions.emplace(Address + It->Rva);
+    const auto* VolatileAccessTableBegin = reinterpret_cast<IMAGE_VOLATILE_RVA_METADATA*>(Address + VolatileMetadata->VolatileAccessTable);
+    const auto* VolatileAccessTableEnd =
+      VolatileAccessTableBegin + (VolatileMetadata->VolatileAccessTableSize / sizeof(IMAGE_VOLATILE_RVA_METADATA));
+    for (auto It = VolatileAccessTableBegin; It != VolatileAccessTableEnd; It++) {
+      VolatileInstructions.emplace(Address + It->Rva);
+    }
+
+    const auto* VolatileInfoRangeTableBegin = reinterpret_cast<IMAGE_VOLATILE_RANGE_METADATA*>(Address + VolatileMetadata->VolatileInfoRangeTable);
+    const auto* VolatileInfoRangeTableEnd =
+      VolatileInfoRangeTableBegin + (VolatileMetadata->VolatileInfoRangeTableSize / sizeof(IMAGE_VOLATILE_RANGE_METADATA));
+    for (auto It = VolatileInfoRangeTableBegin; It != VolatileInfoRangeTableEnd; It++) {
+      VolatileValidRanges.Insert({Address + It->Rva, Address + It->Rva + It->Size});
+    }
   }
 
-  FEXCore::IntervalList<uint64_t> VolatileValidRanges;
-  const auto* VolatileInfoRangeTableBegin = reinterpret_cast<IMAGE_VOLATILE_RANGE_METADATA*>(Address + VolatileMetadata->VolatileInfoRangeTable);
-  const auto* VolatileInfoRangeTableEnd =
-    VolatileInfoRangeTableBegin + (VolatileMetadata->VolatileInfoRangeTableSize / sizeof(IMAGE_VOLATILE_RANGE_METADATA));
-  for (auto It = VolatileInfoRangeTableBegin; It != VolatileInfoRangeTableEnd; It++) {
-    VolatileValidRanges.Insert({Address + It->Rva, Address + It->Rva + It->Size});
+no_file_metadata:
```

Comment:
Split this out into a common function to avoid the goto, think it's also best to print a message when it's active.

LoadImageVolatileMetadata(..)
LoadFEXExtendedVolatileMetadata(...)

AddForceTSOInformation